state/factory/statedb.go

```diff
@@ -82,22 +85,39 @@ func DisableWorkingSetCacheOption() StateDBOption {
 	}
 }
 
+// MetadataNamespaceOption specifies the metadat namespace for versioned DB
+func MetadataNamespaceOption(ns string) StateDBOption {
+	return func(sdb *stateDB, cfg *Config) error {
+		sdb.metaNS = ns
+		return nil
+	}
+}
+
 // NewStateDB creates a new state db
 func NewStateDB(cfg Config, dao db.KVStore, opts ...StateDBOption) (Factory, error) {
 	sdb := stateDB{
 		cfg:                cfg,
 		currentChainHeight: 0,
+		versioned:          cfg.Chain.EnableArchiveMode,
 		registry:           protocol.NewRegistry(),
 		protocolView:       protocol.View{},
 		workingsets:        cache.NewThreadSafeLruCache(int(cfg.Chain.WorkingSetCacheSize)),
-		dao:                dao,
 	}
 	for _, opt := range opts {
 		if err := opt(&sdb, &cfg); err != nil {
 			log.S().Errorf("Failed to execute state factory creation option %p: %v", opt, err)
 			return nil, err
 		}
 	}
+	if sdb.versioned {
+		daoVersioned, ok := dao.(db.KvVersioned)
+		if !ok {
+			return nil, errors.Wrap(ErrNotSupported, "cannot enable archive mode StateDB with non-versioned DB")
+		}
+		sdb.daoVersioned = daoVersioned
+	} else {
+		sdb.dao = dao
+	}
```

Comment:
or we can add a new struct file to implement statedb for archive-mode, like
```
struct stateDBArchive {
    *stateDB
    xxx
}
```
but after checking, that would entail duplicating almost all the codes, and in the future, any changes would need to go to 2 files. Let me know if you have good suggestion.